src/arch/src/aarch64/gic/dist_regs.rs

```diff
@@ -0,0 +1,206 @@
+        // The first part of a shared-irq type of register = (i.e GICD_<REG><0>) is RAZ/WI, so we compute the
+        // base by shifting it with "REG_SIZE * self.bpi" bytes.
+        let base = self.base + REG_SIZE as u32 * self.bpi as u32;
+        let mut end = base;
+        let num_irq = get_num_interrupts(fd)?;
+
+        if self.length > 0 {
+            // This is the single type register (i.e one that is not DIST_X<n>) and for which
+            // the bpi is 0.
+            // Look in the kernel for REGISTER_DESC_WITH_LENGTH.
+            end = base + self.length as u32;
+        }
+        if self.bpi > 0 {
+            // This is the type of register that takes into account the number of interrupts
+            // that the model has. It is also the type of register where
+            // a register relates to multiple interrupts.
+            end = base + (self.bpi as u32 * (num_irq - super::super::layout::IRQ_BASE) / 8);
+            if self.bpi as u32 * (num_irq - super::super::layout::IRQ_BASE) % 8 > 0 {
+                end += REG_SIZE as u32;
+            }
+        }
+        Ok((base, end))
+    }
+
+    #[allow(clippy::trivially_copy_pass_by_ref)]
+    fn access_dist_regs(
+        &self,
+        fd: &DeviceFd,
+        offset: u32,
+        val: &u32,
+        action: &Action,
+    ) -> Result<()> {
+        let mut gic_dist_attr = kvm_bindings::kvm_device_attr {
+            group: kvm_bindings::KVM_DEV_ARM_VGIC_GRP_DIST_REGS,
+            attr: offset as u64,
+            addr: val as *const u32 as u64,
+            flags: 0,
+        };
+        match action {
+            Action::Set => fd.set_device_attr(&gic_dist_attr).map_err(|e| {
+                Error::DeviceAttribute(e, true, kvm_bindings::KVM_DEV_ARM_VGIC_GRP_DIST_REGS)
+            }),
+            Action::Get => fd.get_device_attr(&mut gic_dist_attr).map_err(|e| {
+                Error::DeviceAttribute(e, false, kvm_bindings::KVM_DEV_ARM_VGIC_GRP_DIST_REGS)
+            }),
+        }
+    }
+}
+
+fn get_num_interrupts(fd: &DeviceFd) -> Result<u32> {
+    let num_irqs = 0;
+
+    let mut nr_irqs_attr = kvm_bindings::kvm_device_attr {
+        group: kvm_bindings::KVM_DEV_ARM_VGIC_GRP_NR_IRQS,
+        attr: 0,
+        addr: &num_irqs as *const u32 as u64,
+        flags: 0,
+    };
+    fd.get_device_attr(&mut nr_irqs_attr).map_err(|e| {
+        Error::DeviceAttribute(e, false, kvm_bindings::KVM_DEV_ARM_VGIC_GRP_NR_IRQS)
+    })?;
+    Ok(num_irqs)
+}
+
+/// Get distributor registers of the GIC.
+pub fn get_dist_regs(fd: &DeviceFd) -> Result<Vec<u32>> {
+    let mut state = Vec::new();
+
+    for dreg in VGIC_DIST_REGS {
+        let (mut base, end) = dreg.compute_reg_range(fd)?;
+        while base < end {
+            let val: u32 = 0;
+            dreg.access_dist_regs(fd, base, &val, &Action::Get)?;
+            state.push(val);
+            base += REG_SIZE as u32;
+        }
+    }
```

Comment:
Since `get_dist_regs` and `set_dist_regs` are very similar, can we use the same strategy that is used for redist regs and ICC and create a common method that is called both by `get_dist_regs` and `set_dist_regs` ?